Charts with labels

cryogenian opened the pull request:
I've spent a couple of days trying to generalize charts. And here are wip with labels and generalization
+ `Axes` uses `Set` instead of `Array` --> less time to check if axis is measure
+ I'm removing `Selection` queries and place there `Projection`. That's existentially hidden lens from `State` to `Maybe (D.Dimension Void JCursor)`
+ All states of setup charts now uses `StrMap (D.Dimension Void JCursor)`. This can be hidden too. 
+ To provide interface for this `StrMap` I added `allFields`: `Array Projection` 
+ Chart queries can be generalized and to form `data Query m a = Misc (m a) | OnField Projection (Select a)` where `Select a` is core operation on `dimensionButton` 

Looks like there is a component that should be parameterized by 
```purescript
type Package = 
  { save :: _ 
  , load :: _ 
  , update :: _ 
  }
```
having state `StrMap (Dimension Void JCursor)` and `data Query a = OnField Projection (Select a)`

Also, I think we should remove `SlamData.Form.Select` at all or change it to something like `forall f g. Foldable f => Foldable g => Tuple (f a) (g a)` and use concrete implementation with `Maybe` `Array`. 

That `Reflection a` datatype is to reflect `at` ~ like `at foo == Just foo`, it can be used for extracting `Projection` field. 

Another thing I'm a bit stucked is it possible to generalize `Model` for chart setups. Basically all of them are pretty similar and differ only by labels. 

@natefaubion What do you think? Should we extract this common component handling only `Dimension`s and try to generalize setups model? 

One thing I removed (I hope temporary) is autoselecting. I.e. measure dimension won't be autoselected if there is only one measure dimension.

cryogenian commented:
E.g. having `StrMap LabeledJCursor` as a model could help with autoselecting/loading/saving just because we then have 1-1 map from state and state is constrained by additional funcs.

natefaubion commented on src/SlamData/Workspace/Card/Setups/Common/State.purs:
```diff
@@ -0,0 +1,288 @@
+pack = unsafeCoerce
+
+unpack ∷ ∀ m a b. Projection → ProjectionU m a b
+unpack = unsafeCoerce
```
Is this safe without the polymorphic type annotation at the use site? Seems like you might need some additional newtypery to enforce the `forall a` which is used pervasively. Does this need to be totally generic. Since I think this always works over `StrMap`, maybe it would be easier/safer to go ahead and make it monomorphic?

cryogenian replied:
It work on Reflection too

natefaubion commented:
I really would like a more generic approach to these forms, because updating them is a pretty brutal affair. The main concern I have is around type safety, and while what we have has a lot of boilerplate, all the models _are_ well typed and can't go wrong. I'd like to see if it's possible to keep as much safety as possible. Maybe we can implement something like the type safe HTML DSL with phantom types, if the StrMap approach pays off with boilerplate reduction.

cryogenian commented:
This one is going to be bigger and bigger. 

Basically what's done here 
+ `Axes` uses `Set` instead of `Array`
+ Added component that handles all dimensions for chart setups 
+ This component is parameterized by `Package AnyCardModel (Set.Set JCursor)` 
+ This `Package` can be build (except Parallel) using free monad. The interpreter handles order of buttons and their filtering and so on.

Things that need to be added.
+ Some polishing, triple check that everything works as expected. 
+ Handle `load/save`, it's unsafe :cry: I think it's possible to monomorphise `ModelLens` and `Projection` to use only `Dimension`, and then use coerce for printing and indexing. 
+ Remove some code duplication. (E.g. in child slot modules and traversal compositions) 
+ Actually handle labels and transformation. 
+ Handle previous `aggregationSelectWithNone` (very straightforward thing) 
+ Probably code can be reduced further with row constraints, but stack stopped working on my machine couple days ago and I didn't play with new PS. 

One thing that looks suspicious though is that everything is nicely composed with `Maybe`s in `Setups.Chart.Foo.Eval` --> I'm thinking that we probably could switch to `{dimMap :: StrMap Dimension | r}` in models too. 

Because handling labels much easier with server side aggregation and purescript-sqlsquare I'd rather merge it as is or create separate branch and push sqlsquare and side aggregation changes there. This way prs could be at least 1kLOC size. 

@natefaubion @garyb Could you please take a look at this? What do you think about separate branch?

natefaubion commented on src/SlamData/Workspace/Card/Setups/DimensionMap/Component.purs:
```diff
@@ -0,0 +1,177 @@
+    , receiver: const Nothing
+    }
+  where
+  package = ST.interpret prsm dsl
```
Why don't we pass this directly to the component? It seems like it could save some cycles? Right now I think it'd be interpreting it on every render of the parent.

cryogenian replied:
Mostly because I wanted to hide interpreter from parent component :) I'll replace it with `component'`

cryogenian commented:
@natefaubion Thanks for suggestions this looks much better and more type safe now! 
+ Removed `optional` combinator in favour of `optional` constructor
+ Replaced `AnyLens m` with `Either (ALens' m (LabeledJCursor)) (ALens' m (Maybe LabeledJCursor))` -- loading and saving are typesafe now
+ Made `DimensionMap.Component.component` taking `Package AnyCardModel (Set JCursor)`

cryogenian commented:
@natefaubion Is this good to go? Merging? Branching?

natefaubion commented:
I think the changes are in a much better place, so branching might be in order if there's more followup wrt to SQL^2 stuff and all that.

cryogenian commented:
I've made branch __existential_charts__ 
Closing this

natefaubion commented:
There appears to be an issue where once a dimension is selected, it deselects all the other dimensions. This makes it impossible to actually configure a chart since it keeps getting reset.
